stock_warehouse_clean_delete/models/stock.py

```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
+# © 2016 Akretion (http://www.akretion.com)
+# Sébastien BEAU <sebastien.beau@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openerp import models
+
+
+class AbstractUnlink(models.AbstractModel):
+    _name = "abstract.unlink"
+    _unlink_before = None
+    _unlink_after = None
+
+    def unlink(self):
+        for record in self:
+            for field_to_delete in self._unlink_before:
+                record[field_to_delete].unlink()
+        record_to_delete = set()
+        for record in self:
+            for field_to_delete in self._unlink_after:
+                record_to_delete.add(record[field_to_delete])
+        res = super(AbstractUnlink, self).unlink()
+        for record in record_to_delete:
+            record.unlink()
+        return res
+
+
+class StockWarehouse(models.Model):
+    _inherit = ["stock.warehouse", "abstract.unlink"]
+    _name = "stock.warehouse"
+
+    _unlink_before = [
+        'pick_type_id',
+        'pack_type_id',
+        'out_type_id',
+        'in_type_id',
+        'int_type_id',
+    ]
+
+    _unlink_after = [
+        'view_location_id',
+        'lot_stock_id',
+        'wh_input_stock_loc_id',
+        'wh_qc_stock_loc_id',
+        'wh_output_stock_loc_id',
+        'wh_pack_stock_loc_id',
+        'crossdock_route_id',
+        'reception_route_id',
+        'delivery_route_id',
+        ]
```

Comment:
I do not think that we need to inherit a lot this parts and if we need we can simply inherit the __init__() method to add some field in the list, what do you think?